memory/allocator.go

```diff
@@ -72,6 +65,17 @@ type ResourceAllocator struct {
 	Allocator memory.Allocator
 }
 
+func NewResourceAllocator(allocator memory.Allocator) *ResourceAllocator {
+	// Avoid nesting multiple ResourceAllocator
```

Comment:
It's actually fine if we nest these. This could create an unintended effect. If I write this code:

```
mem := memory.NewResourceAllocator(nestedMem)
fmt.Println(mem.TotalAllocated())
```

That should absolutely print zero.